Add RFC for manifest toml format

toddself opened the pull request:
There are a few things that we likely can add to this, but this covers
the format as it is along with an "entrypoint" specification which we
need currently!

chrisdickinson commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,96 @@
+
+- `"name"` - the canonical name of the package, as a string, including it's namespace. e.g. `"toddself@static-pkg.dev/packge2toml"`
+- `"version"` - a [SEMVER](https://semver.org/) designation conforming to version 2.0.0 of the semantic versioning specification. Due to how TOML parses numbers, this must be represented as a string. e.g. `"1.0.4"`
+- `"entry"` - the filename and package-relative path, as a string, to the main entry point of the package. The root for the relative-path is the directory which contains the `Package.toml` file. e.g `"./src/index.js"`
```
Should `entry` be optional?

chrisdickinson commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,96 @@
+
+## Contributors
+
+- Chris Dickensen [@chrisdickensen](https://github.com/chrisdickensen)
```
s/kensen/kinson/g :smile:

toddself replied:
eek!

chrisdickinson commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,96 @@
+
+The manifest file **MAY** contain zero or more subsections as defined below.
+
+#### `[dependencies]`
```
`devDependencies`, `peerDependencies`, and `optionalDependendies` are also allowed.

toddself commented:
Thanks for the feedback!

olingern commented:
Will **license** be supported inside of a `Package.toml`?

toddself commented:
@ceejbot and @olingern i updated the PR to include all those fields as optional except for `keywords`.  the keywords field has always felt extremely spammy to me and kind of unhelpful.  I'd rather figure out why to better index/weight search results (I've got a lot of ideas about search)

zacanger commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,125 @@
+
+## Introduction
+
+Entropic's cannot share a manifest format with the traditional `package.json` format
```
Entropic's, possessive?

toddself replied:
good catch!

zacanger commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,125 @@
+
+## Filename
+
+The filename of the manifest **MUST** be `Package.toml` and it **MUST** live in
```
There's some chatter about this in #147

zkat replied:
I have serious concerns about this name, especially since [pnpm released support for .yaml and .json5](https://twitter.com/pnpmjs/status/1124733887056424963), and having them support `.toml` is not a stretch. The problems caused by case-normalizing filesystems like HFS+ (with default settings) and NTFS are indeed very serious and they, for example, can seriously mess with tooling like git.

All that said, I understand the Capitalized precedent of so many other modern package managers I can think of, and for that reason, I think having something _like_ this is a good idea.

I'd like to throw my hat in for `Entropic.toml`/`Entropic.lock` as the file pair that this spec uses, if only to distance it _further_ from being a JS-only dependency specifier, but also because a completely different name completely avoids any present or future conflicts in the name style.

lynncyrin replied:
Naming it `[P|p]ackage.toml` sounds like a good call if you're looking to keep CLI api compatibility with npm / yarn / etc. I don't believe that's one of your goals, though?

The first letter of a significant filename needing to be lowercase has made my life easier 100% of the times I've encountered it.

And as someone interested in entropic's model primarily for non-js packages (in the theoretical future where that's supported) - naming is `entropic.toml` (rather than package.toml, presumably for it's just connection) is also a bonus for me.

(aside: let me know if there's some pre-condition for commenting on this PRs, and if so I'll delete this)

zacanger commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,125 @@
+
+- `"entry"` - the filename and package-relative path, as a string, to the main entry point of the package. The root for the relative-path is the directory which contains the `Package.toml` file. e.g `"./src/index.js"`
+- `"type"` - if set to `"module"` this will output a [`package.json` with this included](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field), and will signal to the node interpreter to run this code an ES6 Module.
+- `"license"` - a valid [SPDX license](https://spdx.org/licenses/) for your code
```
Should `license` possibly be required?

toddself replied:
I dislike npm's requirement of it; a lot of code I write is for places I work which necessitates the unwieldy "UNLICENSED" to be typed and like man I typo that thing like a boss

zacanger replied:
Good point, I was thinking mostly of public FOSS packages, not the internal use case.

toddself replied:
also prior art: `license` field in Python's setup.py, Rust's Cargo.toml, Ruby's gemspec all have license as optional

olingern replied:
I think there's a discussion somewhere ( can't remember where ) about a license blocklist, i.e. _our project only allows MIT packages_

Personally, I think licensing is a super important part of installing a package. It's also a neat metric that you could run over a registry's database

chrisdickinson replied:
We should make the `license` field optional, and on the server we can default it to `UNLICENSED` if not given. (& maybe yell about it in the CLI.)

toddself replied:
I updated the spec to say that it'll default to `UNLICENSED` (which I believe is the tactic github takes on this), and that software implementing the entropic fedeverse may complain about any of these optional fields being missing (especially license)

zacanger commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,125 @@
+- `"license"` - a valid [SPDX license](https://spdx.org/licenses/) for your code
+- `"description"` - a short description of your package
+- `"homepage"` - a URL for your project's homepage
+- `"author"` - who wrote this package
```
Should a `repository` or `source` optional field be included? That's something I find super valuable with other tools, and always miss it when it's not there.

toddself replied:
Added repo.  I thought about saying "other fields may appear here, go wild" but then i realized that we might want to not say that in case we need to reserve a field for a future usage that is incompatible with how people are  using it in the wild

zacanger commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,125 @@
+- `"author"` - who wrote this package
+
+These fields **MAY** or **MAY NOT** be validated by CLI tools or the server-side
+software, depending on the implmentation.
```
`implmentation` -> `implementation`

nschonni commented:
Other reference docs from https://github.com/nodejs/package-maintenance might be useful (under the docs folder and the issue tracker). They've been discussing some of those "best practice" subjects when it comes to licensing, files, and versioning, that might be applicable here

toddself commented:
Also prior art:

* Ruby's [Gemspec](https://guides.rubygems.org/specification-reference/)
* Python's [Setup.py](https://docs.python.org/2/distutils/setupscript.html)
* Cargo's [Cargo.toml](https://doc.rust-lang.org/cargo/reference/manifest.html) (thanks @ceejbot)

dmfay commented:
Maven's [pom.xml](https://maven.apache.org/pom.html) format is a bit verbose and the build lifecycle aspect seems out of scope for Entropic, but it does a lot right with dependency resolution and management imo.

toddself commented:
> pom.xml
_takes long drag of cigarette_
> that's a filename i haven't heard in a long time

zkat commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,128 @@
+
+Dependencies lists contain all the resources that are required for
+this package, in various scenarios. The dependencies **MUST** be declared as `"key" = "value"`
+pairs where the `key` is the package's canonical name (as from the top-level section
```
We have a unique opportunity here to move away from this restriction from the original `package.json` and add optional support for objects with extra metadata. This has long been a thorn on our side, and I'd love to see the injustice corrected.

This could _possibly_ translate to the disappearance of `[devDependencies]` and such, or make those categories optional in favor of, e.g. `"toddself@static-pkg.dev/package2toml" = { version: "1.0.4", dev: true, optional: true }`.

chee replied:
i'm so into in this. leaving the future open for a package manager being able to handle more dependency environments than dev & prod. (for instance, i'd love to be able to specify which dependencies are the `test` dependencies)

toddself replied:
yes! your insights are super helpful! thank you

chrisdickinson replied:
YES this. I'd also like to reserve the ability for applications to "open" a dependency to patch it, a la Cargo:

```toml
[dependencies."chris@registry.neversaw.us/foo".dependencies."lodash"]
patch = "chris@registry.neversaw.us/underscore"
```

zkat commented on docs/RFCs/package-toml.md:
````diff
@@ -0,0 +1,128 @@
+To specify a specific version:
+
+```
+"toddself@static-pkg.dev/package2toml" = "1.0.4"
````
A thought comes to mind that these left-hand-side package names are _incredibly unwieldy_, and we'd be better served with leaving these long specifiers to the right-hand side. For example
```toml
[dependencies]
package2toml = "toddself@static-pkg.dev/package2toml@1.0.4"
```
This would instruct the package manager to make the name available as `package2toml` instead of `LONG@NAME/HERE`.

To riff off my earlier comment about objects, users might also choose to split these sections up:

```toml
[dependencies]
package2toml = { from = "toddself@static-pkg.dev/package2toml", version: "1.0.4" }
```

zkat replied:
another thought:

```toml
"toddself@static-pkg.dev/package2toml" = { version: "1.0.4", alias: "package2toml" }
```

but tbh I like the first example I gave much better.

zkat replied:
Oh and from an implementation point of view, having a full-fledged specifier on the RHS makes it much easier to have this sort of thing coexist with existing specifiers (as in, https://github.com/npm/npm-package-arg)

chrisdickinson replied:
(We've discussing whether or not we want to support github-style deps – right now my M.O. is to only allow git deps when run at application level, for patching dependencies.)

dmfay replied:
The long names are a bit awkward, but doesn't assigning them to an arbitrary alias lead to potential weirdness with transitive dependencies? If I make lodash `_` and someone else makes it `lodash`, what happens when someone depends on both our libraries? Letting dependencies dictate what they're called seems like a sensible standard.

zkat replied:
@chrisdickinson I like what Cargo does on this front: you're not allowed to publish using git deps but they're fine for local usage. Also yes, a patch feature would be nice.

@dmfay the package manager is able to differentiate those based on their full identifier + aliases. This is how the alias feature currently works in npm. I'm not worried about conflicts. The only downside here is that there might be less deduplication in the tree.

dmfay replied:
I hadn't heard of an alias feature in npm! I didn't turn anything up in the docs, but I found [this](https://github.com/npm/npm/issues/2943) on the issue tracker which seems to be a conclusive 'no'. Could you elaborate?

I do think there's more at stake than duplication: it becomes less likely that copy+pasted code will "just work" if dependencies it references don't have to have a consistent name, which sounds more annoying the longer I think about it. And I don't know how it'd play with Symbols.

zkat replied:
Sorry, it's my bad. Aliases are... not really documented, and it's a fairly recent feature.

tl;dr you can do:

```json
"dependencies": {
  "underscore": "npm:lodash@1.2.3"
}
```

and it'll install `lodash` under `node_modules/underscore`.

zkat commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,128 @@
+
+- `[dependencies]` - these are the packages that are required to run this package. They will be installed automatically when this package is required by another package, or when you specifically install the dependencies for this package manually.
+- `[devDependencies]` - these are packages that are required to develop this package. They will be installed only when you install the dependencies for this package manually. If a package depends on the current package, they will **NOT** be installed as part of that dependency graph.
+- `[peerDependencies]` - these packages are rqeuired to run this package, however, they will not be installed and should be additionally depended on or installed by either the parent package or along side this package for development.
```
insert generic whining that we need this for compatibility but I hate peerDeps and would love if we just redesigned them.

toddself replied:
i added peerDeps because it's "supported" but god i hate peer deps so gosh darn much

eladb replied:
Peer dependencies night also be the only possible type if Entropic will be used to manage packages for ecosystems which do not support multiple versions of the same dependency in their closures (most of them)

nschonni commented:
Maybe this is something to work out after the initial draft, but what about runtime/installtime platform dependencies. package.json had/has the old "engines" concept for specifying what the package would run on. I'm wondering if you've thought about being able to control what packages get pulled at install time, EX:
- I'm running on Windows, pull down this other package
- I'm running on Node 8<, so pull down an extra package, otherwise it's not needed

zkat commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,128 @@
+The manifest **MUST** contain the following top-level keys:
+
+- `"name"` - the canonical name of the package, as a string, including it's namespace. e.g. `"toddself@static-pkg.dev/packge2toml"`
+- `"version"` - a [SEMVER](https://semver.org/) designation conforming to version 2.0.0 of the semantic versioning specification. Due to how TOML parses numbers, this must be represented as a string. e.g. `"1.0.4"`
```
See also: https://github.com/npm/read-package-json/issues/3

toddself commented:
@nschonni i think that @zkat's ideas around how to [change the dependencies sections](https://github.com/entropic-dev/entropic/pull/189#discussion_r291767838) directly addresses those concerns. 

i've got band practice tonight but I plan on updating this Request for Comments with, well, all these comments!

eladb commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,128 @@
+should be taken to preserve case when possible.
+
+## Format
+
```
Curious about the decision to use TOML instead of JSON. I’ve seen tremendous value in the ability to parse package.json files “natively” in node. Maybe worth describing the motivation for this decision as I am sure it will be a common question.

cmmartti replied:
Comments, I expect. I've sometimes wanted to add a comment explaining why a dependency is pinned at a specific version, or add a description to a "scripts" command, but JSON doesn't allow that.

You could go with something like [JSON5](https://json5.org) but then you lose the ability to parse it "natively" in Node, and at that point it doesn't really matter what you use.

zacanger replied:
There was some discussion about why TOML in https://github.com/entropic-dev/entropic/issues/106 and https://github.com/entropic-dev/entropic/issues/162, in particular https://github.com/entropic-dev/entropic/issues/106#issuecomment-497982913 and https://github.com/entropic-dev/entropic/issues/162#issuecomment-499264287

cmmartti replied:
Those issues don't really explain why TOML was chosen, only why Entropic doesn't use package.json. Although I support the use of TOML, there's no reason the package file couldn't be entropic.json instead.

toddself replied:
JSON doesn't support comments which has been a frequent complaint about it (especially when used for manifests). I'm also adding a "why TOML" section to the RFC

eladb commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,128 @@
+- `"description"` - a short description of your package
+- `"homepage"` - a URL for your project's homepage
+- `"author"` - who wrote this package
+- `"repository"` - where does this code live
```
Location within the repo for mono-repos would be useful.

Generally mono-repos and large projects should get some attention. For example, being able to reuse package definitions across multiple packages. Happy to provide mode context when it’s relevant.

zkat replied:
👍 . See [the related RFC for npm](https://github.com/npm/rfcs/blob/latest/implemented/0010-monorepo-subdirectory-declaration.md)

toddself commented:
> Maybe this is something to work out after the initial draft, but what about runtime/installtime platform dependencies. package.json had/has the old "engines" concept for specifying what the package would run on. I'm wondering if you've thought about being able to control what packages get pulled at install time, EX:
> 
>     * I'm running on Windows, pull down this other package
> 
>     * I'm running on Node 8<, so pull down an extra package, otherwise it's not needed

I think we can handle this with the expanded object scope

toddself commented:
Link to discourse discussion: https://discourse.entropic.dev/t/the-toml-file-lets-talk-about-it/23/17

toddself commented:
So far I've resolved all the conversations in this PR except:

* what to call the file itself
* peerDependencies

Please make new comments if I haven't captured your intent correctly! (There was a _lot_ of **excellent, positive** feedback!!!)

zacanger commented:
Maybe peerDeps could just be a future conversation and possible feature?

toddself commented:
A few more thoughts!

I think once we decide on the name for the file (which seems to be leaning towards`entropic.toml` and `lock.entropic` I think we can punt on the `peerDeps` issue until later, so ideally we can get this merged in and start building support for all these wonderful features

toddself commented:
Looks like @zacanger was 42 seconds faster with this thought!

toddself commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+- `"license"` - a valid [SPDX license](https://spdx.org/licenses/) for your code. If a license is not provided then the code is considered to be `UNLICENSED`.
+- `"description"` - a short description of your package
+- `"homepage"` - a URL for your project's homepage
+- `"author"` - who wrote this package
```
so i punted on this a little and just said "valid email identifiers or an array of them"

toddself commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+
+- `"entry"` - the filename and package-relative path, as a string, to the main entry point of the package. The root for the relative-path is the directory which contains the `[TK].toml` file. e.g `"./src/index.js"`.
+- `"type"` - if set to `"module"` this will output a [`package.json` with this included](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field), and will signal to the node interpreter to run this code an ES6 Module.
+- `"license"` - a valid [SPDX license](https://spdx.org/licenses/) for your code. If a license is not provided then the code is considered to be `UNLICENSED`.
```
this is literally the source of this.  the number of bullshit descriptions i've added  to packages...

zkat commented:
I think it's worth acknowledging that wherever this lands, it'll only be a starting point for more earnest development of a manifest spec that evolves as the client is developed. We can probably expect a lot of changes, so my inclination is to not get too hung up on little details, but probably just settle on the name and the basic dep specification stuff and otherwise this is a v good doc to start from <3

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+associated with the ECMAScript ecosystem for a number of reasons:
+
+1. Entropic is intended to allow for federated package management for more than just the ECMAScript ecosystem
+1. The `package.json` name specifier does not allow for the namespacing that entropic provides
```
can you expand on why this is the case?

toddself replied:
We can't support some of the options that we'd like to while keeping with a legacy format. in addition certain valid specifiers for entropic are [not valid for legacy registry clients](https://github.com/entropic-dev/entropic/issues/106#issuecomment-497982913)

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+
+## Filename
+
+The filename of the manifest **MUST** be `[TK].toml` and it **MUST** live in
```
Can a package have multiple roots? That is currently the case for projects in the Node.js ecosystem where it is possible to have nested projects

toddself replied:
Yes, there will be support for monorepo (ncluding the location of this package within the monorepo).  

This should work like it does with the existing package manager.

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+
+The manifest **MUST** contain the following top-level keys, in order to be published:
+
+- `"name"` - the canonical name of the package, as a string, including its namespace. e.g. `"toddself@static-pkg.dev/packge2toml"`
```
is the idea that the canonical name of imported packages will also include the namespace?

e.g. `import 'toddself@static-pkg.dev/packge2toml'`

toddself replied:
IIRC yes, but @chrisdickinson or @ceejbot might have a better answer for this.

chrisdickinson replied:
Yes. Additionally, I was thinking about making packages available by the "smallest unique" name at each level too – so if there's no other `package2toml` at a given level, you could require it via `require('package2toml')`. However, that comes with a cost – so we deferred that behavior until later.

zkat replied:
requiring using the fully qualified name feels like a big UX fail, imo. I think there can/should be some rule enforcement that makes it the default -- see also Mastodon's `@foo` being context-aware, so you don't have to type out someone's fully-qualified name pretty much -ever-.

olingern replied:
Maybe it's possible for us to only require the package name by symlink'ing the downloaded package into the`node_modules` folder. 

So, `ds` would install `toddself@static-pkg.dev/packge2toml` as `package2toml`. I think this becomes more complex and simply importing the fully qualified name seems less error prone.

zkat replied:
@olingern aliases solve this use case nicely

chrisdickinson replied:
@zkat:

> requiring using the fully qualified name feels like a big UX fail, imo. I think there can/should be some rule enforcement that makes it the default -- see also Mastodon's `@foo` being context-aware, so you don't have to type out someone's fully-qualified name pretty much -ever-.

Yep, agreed – I want this feature (and aliases!) At the time, I deferred implementing the context-aware requirement until later because when I went to implement it, it seemed like it would have performance and deduping consquences that I wasn't prepared to tackle at the time. (Also: interactions with how legacy packages are unpacked vs. entropic packages.)

My goal was to produce a working system from which we could iterate. It seemed at the time like we could defer aliasing and context-aware import support until later, while reserving the right to re-define our existing dependency item format as a shorthand for an expanded one that supports those features down the line. (I was starting to get the "I'm too far out on a limb without supports" feeling from speccing aliasing and context-awareness without working from a backing implementation, if that makes sense?)

zkat replied:
makes total sense to me :)

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+
+The manifest **MAY** contain the following fields:
+
+- `"entry"` - the filename and package-relative path, as a string, to the main entry point of the package. The root for the relative-path is the directory which contains the `[TK].toml` file. e.g `"./src/index.js"`.
```
is there a reason to stray from `main` which has canonically been used in the node.js ecosystem? This seems like it would create an extra source of confusion especially if it needs to be translated to main in a toml -> json conversion

toddself replied:
good point!

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+The manifest **MAY** contain the following fields:
+
+- `"entry"` - the filename and package-relative path, as a string, to the main entry point of the package. The root for the relative-path is the directory which contains the `[TK].toml` file. e.g `"./src/index.js"`.
+- `"type"` - if set to `"module"` this will output a [`package.json` with this included](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field), and will signal to the node interpreter to run this code an ES6 Module.
```
the type doesn't actually signal to the loader what type to load, it signifies how to interpret file extensions

You can have `pkg.type` unset, and it would default to `commonjs`. Based on this current behavior having no `type` field and a `main` with a `.mjs` extension would still export an esm module.

Similarly a `type` of `module` with a `main` with the `.cjs` extension will still export a cjs module

toddself replied:
clarified!

MylesBorins commented on docs/RFCs/package-toml.md:
```diff
@@ -0,0 +1,166 @@
+- `[dependencies]` - these are the packages that are required to run this package. They will be installed automatically when this package is required by another package, or when you specifically install the dependencies for this package manually.
+- `[devDependencies]` - these are packages that are required to develop this package. They will be installed only when you install the dependencies for this package manually. If a package depends on the current package, they will **NOT** be installed as part of that dependency graph.
+- `[peerDependencies]` - these packages are rqeuired to run this package, however, they will not be installed and should be additionally depended on or installed by either the parent package or along side this package for development.
+- `[optionalDependencies]` - these packages are not required to run this software, but may provide additional options or features in this package. They must be required or installed by the parent package or along side this package for development.
```
I've heard requests for better ways of handling optional dependencies. E.g. how do I not install / compile brotli on newer versions of node.

toddself replied:
Yeah, i think we have some definite work to do here in the deps section, especially with regard to the notably absent lifecycle scripts...

chrisdickinson commented on docs/RFCs/001-manifest-toml.md:
```diff
@@ -0,0 +1,166 @@
+
+#### Dependency Lists
+
+Dependencies lists contain all the resources that are required for this package, in various scenarios. Dependencies consist of a `key/value` pair where the `key` is a string that provides a unique name for the package in this project. This `key` can be either the qualified package name from the package's manifest, or an <a href="#alias">alias</a> for the package.
```
(Ah! And I just remembered: there is a shorthand syntax at play for keys: if you provide just `"lodash"`, that'll expand to `"legacy@<current registry>/lodash`)

chrisdickinson commented:
We'll treat this as a living document & continue to iterate and fill things in as we get to them, but this is a great place to start from: THANK YOU!

static-raccoon commented:
This pull request has been mentioned on **Entropic**. There might be relevant details there:

https://discourse.entropic.dev/t/the-toml-file-lets-talk-about-it/23/1

static-raccoon commented:
This pull request has been mentioned on **Entropic**. There might be relevant details there:

https://discourse.entropic.dev/t/lifecycle-scripts-and-native-binaries/73/1
